### ISSUE TITLE
Formatting curried functions should have a better heuristic for breaking

### ISSUE THREAD

**Issue by jsnajdr** [prettier/prettier#12679]
**Prettier 2.6.2**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEYCuAnPATwEZMBeTACgEMBKMgPkoCN6HqXaBuAHSkigZs+IgCZyVOo05JqUQow6te-dFlwFCAZgk02AHgAqTVos4qBQjUQAsuqU2DMAvmyXc+IADQgIABxgAS3RkUGoCCAB3AAVwhDRkEGoAG0jqQgSfZjxqMABrOBgAZWoAWzgAGUCoOGQAMxS0OCyc-MKiv1zqgHNkGDwcZpA4UuY4ABNxiYq5bpxqbrgAMQg8UuoYIKhelGocGAhvEAALGFLkgHVjwPg0TrA4IvibwIA3G8JEsDRMkGqmvAwaI5brreqNIYAKzQAA8ij1knAAIo4CDwcHJJo+Tp4AGJZjUMbJI5+PDVGAXQLjGDHZDEAAM9OxeAgTQuOT8iVJcABr1qPgAjqj4MD-AldmgALQ1CYTI54OBCwIK4ELMFIBqYoZNUqBPoDbUI5HC2oaiE+GCEynU2lIUQWnKBZI9ADCEFK6uGaAArEccE1DITxZqsSBXoMAJJQKawIpgMkBACC0aKMEIiIxTWcziAA)

```sh
--parser babel
--print-width 100
```

**Input:**

```jsx
const curry1 = (a)=>(b)=>a(b);
const curry2 = (a)=>(b):any=>a(b);
const curry3 = (a)=><T>(b)=>a(b);
const curry4 = (a)=>({b})=>a(b);
const curry5 = (a)=>(b=1)=>a(b);
```

**Output:**

```jsx
const curry1 = (a) => (b) => a(b);
const curry2 =
  (a) =>
  (b): any =>
    a(b);
const curry3 =
  (a) =>
  <T>(b) =>
    a(b);
const curry4 =
  (a) =>
  ({ b }) =>
    a(b);
const curry5 =
  (a) =>
  (b = 1) =>
    a(b);
```

**Expected behavior:**
I'd expect that all five functions are formatted as one-liners, just like in the input. The [heuristic that triggers the breaking](https://github.com/prettier/prettier/blob/main/src/language-js/print/function.js#L303-L308) does that way too often: whenever there's a return type, type parameter, destructuring, default value... There's no good reason for that.

#9992 originally intended to fix #9985, where one of the function signatures got broken into several lines:
```jsx
const curry =
  (a) =>
  (b) =>
  ({
    c: d,
    e: f,
  }) =>
  (g) =>
    a(b(d + f));
```
Here it's very reasonable to break the curry chain into several lines, too, instead of doing the original:
```jsx
const curry = (a) => (b) => ({
  c: d,
  e: f,
}) => (g) => a(b(d + f));
```
But there's never a good reason to break the curry chain if it could fit on one line.

Another reason why the decision to break the curried arguments should have a higher threshold is that it changes indentation of the function body. A simple argument change can produce a very large diff. That should be avoided. Prettier authors followed that principle when merging #11515, which changes formatting of complex types only to avoid changing the body indentation.

I'm aware that my issue is a duplicate of #11367, but I think it was closed too bluntly and that it deserves more thoughtful consideration.

**Comment by wamasimba** [comment]
As I'm leaning more and more on curried functions to create easily readable code, I also ran into this; AFAIK even the latest prettier (currently using 2.8.6) messes up my clean code code into illegible lines of semi-statements.

More specifically, I want to create many utilities like:
```ts
const withType = (desired: string) => ({ type }) => type === desired;
```

To be used in code like:
```ts
const value = data.find(withType('example'))?.value;
```

Hoever, prettier garbles that utility into:
```ts
const withType =
    (desired: string) =>
    ({ type }) =>
        type === desired;
```
Which, IMNSHO is the least readable version I could think of....

**Comment by nmsmith** [comment]
I'm encountering this issue too. In particular, functions that accept a `...` parameter seem to cause line-breaking for some inexplicable reason:

```
   let foo = (pattern: Pattern) => () => {}
   let bar =
      (...patterns: Pattern[]) =>
      () => {}
```

**Comment by gyerzh** [comment]
My team also encounters this issue, particularly since we use such curried functions extensively. After formatting, the code becomes less readable.

**Input:**
`export const goTo = ({ obj })  => param => () => obj.doSomething(param)`

**Output:**
```
export const goTo =
  ({ obj }) =>
  param =>
  () =>
    obj.doSomething(param)
```


**Expected behavior:**
`export const goTo = ({ obj })  => param => () => obj.doSomething(param)`

**Comment by tstewart-klaudhaus** [comment]
Agree, I'm having to use `// prettier-ignore` in order to maintain readable code in these cases.
